WIP: Upgrade fotg210 drivers to latest Faraday provided drivers

grant-olson opened the pull request:
This is a work in progress to get USB working on the Pine64 Ox64. The USB IP seems to be based on the fotg210 from Faraday Semiconductors. Attempting to enable the mainline drivers had all sorts of problems.

This attempts to use the faraday drivers at https://github.com/FaradayA380Platform/Linux.git with mixed, but better, results.

Current Status

HCD Host Driver:
* Loads up but hangs when you plug in a good device.
* Has same problem is bl_mcu_sdk examples where devices using BULK transfer modes don't mount properly.

UDC Device Status:
* Seems to load fine.
* When mounting a gadget, such as `modprobe g_serial` we get a segfault.
* Surprisingly the segfault doesn't impact the mapping between the FOTG210 and Host Device, things show up on host with `lsusb`
* Attempting to use gadgets fail, presumably because the segfault stops the gadget modules from loading.

grant-olson committed:
Delete mainline fotg210 drivers and replace with drivers from Faraday linux fork. This commit is all Faraday only code, no patches.

grant-olson committed:
Patches to make faraday fork fotg210 drivers work on linux 6.2

grant-olson commented:
For this code to work you will need to power up the USB device in low level code. I am doing that with this fork of OBLFR xxx

grant-olson commented:
Actually this branch https://github.com/grant-olson/OBLFR/tree/usbphy/components

Hunter1753 commented:
Hey, I am very interested in kernel driver development/ implementation for the bl808. Is there any way I could help you with testing this?

grant-olson commented:
Got a mass storage device starting to work. This is easier to see what's going on than the network or serial gadgets. Here is how I set up the filesystem for the mass storage device:

Make an empty file for storage:
```
dd bs=1M count=32 if=/dev/zero of=/backing_file
```

Use fdisk to paritition:
* `fdisk -C 32 -H 16 -S 128 /backing_file`
* type **n** to create new partition.
* type **p** to select primary partition.
* type **1** for partition 1.
* Keep defaults for start and end of partition.
* type **t** to set the partition type.
* type **b** to do FAT32.
* *type **w** to write new partition table and quit.

Format the partition `mkfs.vfat -n OPENBFL /backing_file`

Mount the partition to add files:
* `mkdir /tmp/bak`
* `mount /backing_file /tmp/bak -t vfat`
* `nano README.html` or something else to add some files.
* `umount /tmp/bak`

Creating the mass storage device: `modprobe g_mass_storage file=/backing_file ro=y removable=y`

Then plug the USB C adapter in to a host computer.

grant-olson commented:
@Hunter1753 not sure what your skill level is, but now I think the most useful way to help would be to build the code and try to bring up different usb gadgets and properly configure them. For example I can bring up the `g_ether` module and both my host system and the Ox64 think there is a network, but I haven't figured out how to issue the commands so traffic routes between them. Similar problem with `g_serial` where it says there's a serial monitor but I can't bring up a linux shell.

But it's a little touchy now since you need to build the kernel from scratch. Once things are a little more stable I hope to get an updated buildroot image up this week to make it easier for others to bring up the drivers and tinker around. I'll comment here when I do.

grant-olson committed:
Update device tree to sync with arm000 tree

grant-olson committed:
New device tree entries for fotg HCD and UDC drivers.

grant-olson commented:
I've rebased against the latest linux RC and squashed down some of the commits to make the code cleaner.
I've also set up a bouffalo_buildroot branch and release that makes it easy to get this code on a machine. The release has all the binaries to flash, and the branch should make it easier to mess around with various gadgets.

https://github.com/grant-olson/buildroot_bouffalo/releases

grant-olson committed:
Fix for runtime error when plugging in USB objects

Hunter1753 commented:
@grant-olson i see. Yes, i am very new to this all, so i have to figure out some stuff, but with the release of you, i will try a bit
